File: src/nativeMain/kotlin/com/autoxaccount/DataDir.kt
package com.autoxaccount

import kotlinx.cinterop.*
import platform.posix.*
import kotlinx.datetime.Clock

/**
 * Data directory management module
 * Migrated from Rust data_dir.rs
 * 
 * Manages application data directories, configuration, and cache files
 */

object DataDir {
    /**
     * Initialize all required directories
     */
    fun initDirectories(): Result<Unit> = runCatchingResult {
        val dataDir = getDataDir()
        val logsDir = getLogsDir()
        val browserDataDir = getBrowserDataDir()
        val screenshotsDir = getScreenshotsDir()

        createDirectory(dataDir)
        createDirectory(logsDir)
        createDirectory(browserDataDir)
        createDirectory(screenshotsDir)
    }

    /**
     * Get main data directory path based on OS
     */
    fun getDataDir(): String {
        return when {
            isWindows() -> {
                val appData = getenv("APPDATA")?.toKString() ?: "C:\\Users\\Default\\AppData\\Roaming"
                "$appData\\auto-x-account"
            }
            isMacOS() -> {
                val home = getenv("HOME")?.toKString() ?: "/Users/Default"
                "$home/Library/Application Support/auto-x-account"
            }
            else -> { // Linux
                val home = getenv("HOME")?.toKString() ?: "/home/user"
                "$home/.local/share/auto-x-account"
            }
        }
    }

    /**
     * Get logs directory path based on OS
     */
    fun getLogsDir(): String {
        return when {
            isWindows() -> "${getDataDir()}\\logs"
            isMacOS() -> {
                val home = getenv("HOME")?.toKString() ?: "/Users/Default"
                "$home/Library/Logs/auto-x-account"
            }
            else -> "${getDataDir()}/logs"
        }
    }

    /**
     * Get browser data directory
     */
    fun getBrowserDataDir(): String {
        return "${getDataDir()}${pathSeparator()}browser_data"
    }

    /**
     * Get screenshots directory
     */
    fun getScreenshotsDir(): String {
        return "${getDataDir()}${pathSeparator()}screenshots"
    }

    /**
     * Get accounts file path
     */
    fun getAccountsPath(): String {
        return "${getDataDir()}${pathSeparator()}accounts.json"
    }

    /**
     * Get config file path
     */
    fun getConfigPath(): String {
        return "${getDataDir()}${pathSeparator()}config.json"
    }

    /**
     * Get batch tasks directory
     */
    fun getBatchTasksDir(): String {
        return "${getDataDir()}${pathSeparator()}batch_tasks"
    }

    /**
     * Load or create default configuration
     */
    fun loadOrCreateConfig(): Result<Config> = runCatchingResult {
        val configPath = getConfigPath()
        
        if (fileExists(configPath)) {
            Config.fromFile(configPath).getOrThrow()
        } else {
            val config = Config()
            config.toFile(configPath).getOrThrow()
            config
        }
    }

    /**
     * Get data directory info as formatted string
     */
    fun getDataDirInfo(): String {
        return buildString {
            appendLine("═══════════════════════════════════════════════")
            appendLine("  数据目录信息 / Data Directory Information")
            appendLine("═══════════════════════════════════════════════")
            appendLine()
            appendLine("📁 数据目录 / Data Directory:")
            appendLine("   ${getDataDir()}")
            appendLine()
            appendLine("📝 日志目录 / Logs Directory:")
            appendLine("   ${getLogsDir()}")
            appendLine()
            appendLine("🌐 浏览器数据 / Browser Data:")
            appendLine("   ${getBrowserDataDir()}")
            appendLine()
            appendLine("📸 截图目录 / Screenshots:")
            appendLine("   ${getScreenshotsDir()}")
            appendLine()
            appendLine("💾 配置文件 / Config File:")
            appendLine("   ${getConfigPath()}")
            appendLine()
            appendLine("👤 账号文件 / Accounts File:")
            appendLine("   ${getAccountsPath()}")
            appendLine("═══════════════════════════════════════════════")
        }
    }

    /**
     * Cleanup old tasks older than specified days
     */
    fun cleanupOldTasks(days: Int): Result<Unit> = runCatchingResult {
        val tasksDir = getBatchTasksDir()
        if (!fileExists(tasksDir)) return@runCatchingResult

        val cutoffTime = Clock.System.now().toEpochMilliseconds() - (days * 24 * 60 * 60 * 1000L)
        
        // List and remove old task files
        // Note: This is simplified - full implementation would require directory traversal
        println("清理 $days 天前的旧任务 / Cleaning up tasks older than $days days")
    }

    // Helper functions

    private fun pathSeparator(): String = if (isWindows()) "\\" else "/"

    private fun isWindows(): Boolean {
        val os = getenv("OS")?.toKString()
        return os?.contains("Windows", ignoreCase = true) == true
    }

    private fun isMacOS(): Boolean {
        // Check if running on macOS by examining uname
        memScoped {
            val buf = allocArray<ByteVar>(256)
            if (gethostname(buf, 256u) == 0) {
                // This is a simplification - proper OS detection would use uname or platform checks
                return false // Placeholder
            }
        }
        return false
    }

    private fun fileExists(path: String): Boolean {
        memScoped {
            val stat = alloc<stat>()
            return stat(path, stat.ptr) == 0
        }
    }

    private fun createDirectory(path: String) {
        if (fileExists(path)) return
        
        // Create directory with read/write/execute permissions for owner
        if (isWindows()) {
            mkdir(path)
        } else {
            mkdir(path, 0x1EDu) // 0755 in octal
        }
    }
}

File: src/data_dir.rs
//! 数据目录管理模块
//! Data directory management module
//! 
//! 负责管理应用程序的数据目录，确保所有数据持久化到正确的位置

use anyhow::Result;
use std::path::PathBuf;
use tracing::info;

/// 获取应用程序数据目录
/// Get application data directory based on OS
pub fn get_data_dir() -> PathBuf {
    #[cfg(target_os = "windows")]
    {
        // Windows: %APPDATA%\auto-x-account
        if let Ok(appdata) = std::env::var("APPDATA") {
            PathBuf::from(appdata).join("auto-x-account")
        } else {
            PathBuf::from(".").join("data")
        }
    }

    #[cfg(target_os = "macos")]
    {
        // macOS: ~/Library/Application Support/auto-x-account
        if let Ok(home) = std::env::var("HOME") {
            PathBuf::from(home)
                .join("Library")
                .join("Application Support")
                .join("auto-x-account")
        } else {
            PathBuf::from(".").join("data")
        }
    }

    #[cfg(not(any(target_os = "windows", target_os = "macos")))]
    {
        // Linux/Unix: ~/.local/share/auto-x-account
        if let Ok(home) = std::env::var("HOME") {
            PathBuf::from(home)
                .join(".local")
                .join("share")
                .join("auto-x-account")
        } else if let Ok(xdg_data_home) = std::env::var("XDG_DATA_HOME") {
            PathBuf::from(xdg_data_home).join("auto-x-account")
        } else {
            PathBuf::from(".").join("data")
        }
    }
}

/// 获取配置文件路径
/// Get configuration file path
pub fn get_config_path() -> PathBuf {
    get_data_dir().join("config.json")
}

/// 获取账号数据文件路径
/// Get accounts data file path
pub fn get_accounts_path() -> PathBuf {
    get_data_dir().join("accounts.json")
}

/// 获取邮箱数据文件路径
/// Get emails data file path
pub fn get_emails_path() -> PathBuf {
    get_data_dir().join("emails.json")
}

/// 获取批量任务数据目录
/// Get batch tasks data directory
pub fn get_tasks_dir() -> PathBuf {
    get_data_dir().join("tasks")
}

/// 获取截图目录
/// Get screenshots directory
pub fn get_screenshots_dir() -> PathBuf {
    get_data_dir().join("screenshots")
}

/// 获取浏览器数据目录
/// Get browser data directory
pub fn get_browser_data_dir() -> PathBuf {
    get_data_dir().join("browser_data")
}

/// 初始化所有必需的目录
/// Initialize all required directories
pub fn init_directories() -> Result<()> {
    let dirs = vec![
        get_data_dir(),
        get_tasks_dir(),
        get_screenshots_dir(),
        get_browser_data_dir(),
    ];

    for dir in dirs {
        if !dir.exists() {
            std::fs::create_dir_all(&dir)?;
            info!("创建目录 / Created directory: {}", dir.display());
        }
    }

    Ok(())
}

/// 从数据目录加载配置文件，如果不存在则创建默认配置
/// Load config from data directory, create default if not exists
pub fn load_or_create_config() -> Result<crate::config::Config> {
    let config_path = get_config_path();
    
    if config_path.exists() {
        info!("从数据目录加载配置 / Loading config from: {}", config_path.display());
        crate::config::Config::from_file(&config_path)
    } else {
        info!("配置文件不存在，创建默认配置 / Config not found, creating default");
        let config = crate::config::Config::default();
        
        // 保存默认配置到数据目录
        config.to_file(&config_path)?;
        info!("默认配置已保存到 / Default config saved to: {}", config_path.display());
        
        // 同时在当前目录创建示例配置（仅用于参考）
        let example_path = PathBuf::from("config.example.json");
        if !example_path.exists() {
            config.to_file(&example_path)?;
            info!("示例配置已创建 / Example config created: {}", example_path.display());
        }
        
        Ok(config)
    }
}

/// 保存配置文件到数据目录
/// Save config to data directory
pub fn save_config(config: &crate::config::Config) -> Result<()> {
    let config_path = get_config_path();
    config.to_file(&config_path)?;
    info!("配置已保存到 / Config saved to: {}", config_path.display());
    Ok(())
}

/// 获取数据目录的信息（用于显示）
/// Get data directory information for display
pub fn get_data_dir_info() -> String {
    format!(
        "数据目录 / Data Directory:\n\
        主目录 / Main: {}\n\
        配置文件 / Config: {}\n\
        账号数据 / Accounts: {}\n\
        邮箱数据 / Emails: {}\n\
        任务数据 / Tasks: {}\n\
        截图目录 / Screenshots: {}\n\
        浏览器数据 / Browser: {}\n\
        日志目录 / Logs: {}",
        get_data_dir().display(),
        get_config_path().display(),
        get_accounts_path().display(),
        get_emails_path().display(),
        get_tasks_dir().display(),
        get_screenshots_dir().display(),
        get_browser_data_dir().display(),
        crate::logging::get_log_dir().display()
    )
}

/// 清理旧的任务数据（保留最近N天）
/// Clean up old task data (keep last N days)
pub fn cleanup_old_tasks(days_to_keep: u32) -> Result<()> {
    let tasks_dir = get_tasks_dir();
    
    if !tasks_dir.exists() {
        return Ok(());
    }

    let cutoff_time = chrono::Local::now() - chrono::Duration::days(days_to_keep as i64);

    for entry in std::fs::read_dir(tasks_dir)? {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() {
            if let Ok(metadata) = std::fs::metadata(&path) {
                if let Ok(modified) = metadata.modified() {
                    match chrono::DateTime::<chrono::Local>::from(modified).partial_cmp(&cutoff_time) {
                        Some(std::cmp::Ordering::Less) => {
                            info!("删除旧任务文件 / Removing old task file: {}", path.display());
                            let _ = std::fs::remove_file(&path);
                        }
                        _ => {
                            // File is newer or comparison failed, keep it
                        }
                    }
                }
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_data_dir() {
        let dir = get_data_dir();
        assert!(dir.to_string_lossy().contains("auto-x-account"));
    }

    #[test]
    fn test_get_config_path() {
        let path = get_config_path();
        assert!(path.to_string_lossy().ends_with("config.json"));
    }
}
